From 51c9ba0f1b7660dc8fb45fee801eb4360a2eec35 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Olivier=20De=20Canni=C3=A8re?= <olivier.decanniere@qt.io>
Date: Tue, 14 Oct 2025 15:08:38 +0200
Subject: [PATCH 214/239] qmllint: Augment the import warning for unspecialized
 list properties

Pick-to: 6.8
Change-Id: Ifd33df7fc2cfeb86c5af9a9dc752518c52d5b91b
Reviewed-by: Sami Shalayel <sami.shalayel@qt.io>
(cherry picked from commit 4f4b36ba4fd17476f4bba18bbe16e299db408c01)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/qmlcompiler/qqmljsimportvisitor.cpp | 6 ++++--
 tests/auto/qml/qmllint/tst_qmllint.cpp  | 6 ++++++
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/src/qmlcompiler/qqmljsimportvisitor.cpp b/src/qmlcompiler/qqmljsimportvisitor.cpp
index ec6c573ec5..db9d4c5499 100644
--- a/src/qmlcompiler/qqmljsimportvisitor.cpp
+++ b/src/qmlcompiler/qqmljsimportvisitor.cpp
@@ -836,8 +836,10 @@ void QQmlJSImportVisitor::processPropertyTypes()
             property.setType(propertyType);
             type.scope->addOwnProperty(property);
         } else {
-            m_logger->log(property.typeName() + ' '_L1 + wasNotFound + ' '_L1 + didYouAddAllImports,
-                          qmlImport, type.location);
+            QString msg = property.typeName() + ' '_L1 + wasNotFound + ' '_L1 + didYouAddAllImports;
+            if (property.typeName() == "list"_L1)
+                msg += " list is not a type. It requires an element type argument (eg. list<int>)"_L1;
+            m_logger->log(msg, qmlImport, type.location);
         }
     }
 }
diff --git a/tests/auto/qml/qmllint/tst_qmllint.cpp b/tests/auto/qml/qmllint/tst_qmllint.cpp
index f284e15fe3..79f2b4f6a5 100644
--- a/tests/auto/qml/qmllint/tst_qmllint.cpp
+++ b/tests/auto/qml/qmllint/tst_qmllint.cpp
@@ -1496,6 +1496,12 @@ void TestQmllint::dirtyQmlSnippet_data()
             << u"property var v: {}"_s
             << Result{ { { "Unintentional empty block, use ({}) for empty object literal"_L1, 1, 17 } } }
             << defaultOptions;
+    QTest::newRow("unspecializedList")
+            << u"property list l"_s
+            << Result{ { { "list was not found. Did you add all imports and dependencies? list is "
+                           "not a type. It requires an element type argument (eg. list<int>)"_L1,
+                           1, 1 } } }
+            << defaultOptions;
     QTest::newRow("upperCaseId")
             << u"id: Root"_s
             << Result{ { { "Id must start with a lower case letter or an '_'"_L1, 1, 5 } } }
-- 
2.51.2

